sound/soc/sof/ops.c

```diff
@@ -167,18 +167,21 @@ void snd_sof_dsp_panic(struct snd_sof_dev *sdev, u32 offset, bool non_recoverabl
 			 __func__, sdev->dsp_oops_offset, offset);
 
 	/*
-	 * Only print the panic information if we have non recoverable panic or
-	 * if all dumps should be printed
+	 * Set the fw_state to crashed only in case of non recoverable DSP panic
+	 * event.
+	 * Use different message within the snd_sof_dsp_dbg_dump() depending on
+	 * the non_recoverable flag.
 	 */
-	if (non_recoverable || sof_debug_check_flag(SOF_DBG_PRINT_ALL_DUMPS)) {
-		/* We want to see the DSP panic! */
-		sdev->dbg_dump_printed = false;
-
+	sdev->dbg_dump_printed = false;
+	if (non_recoverable) {
 		snd_sof_dsp_dbg_dump(sdev, "DSP panic!",
 				     SOF_DBG_DUMP_REGS | SOF_DBG_DUMP_MBOX);
-		if (non_recoverable)
-			sof_set_fw_state(sdev, SOF_FW_CRASHED);
+		sof_set_fw_state(sdev, SOF_FW_CRASHED);
 		snd_sof_trace_notify_for_error(sdev);
+	} else {
+		snd_sof_dsp_dbg_dump(sdev,
+				     "DSP panic (recovery will be attempted)",
```

Comment:
snd_sof_dsp_panic() is called when a panic occurs irrespective of whether it is during FW boot or afterwards. There's no recovery in case of a panic when it happens after the FW has succesfully booted. So this change in a common function doesnt make a lot of sense I think

Reply:
@ranj063, this was my understanding as well until @lyakh showed me this log fragment:
<details>
  <summary>Firmware boot spiced up with a DSP panic</summary>

```
[ 2267.902592] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: Turning i915 HDAC power 1
[ 2267.907506] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: Debug PCIR: 00000000 at  00000044
[ 2267.910443] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: Debug PCIR: 00000040 at  00000048
[ 2267.910471] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: Debug PCIW: 00000000 at  00000048
[ 2267.914246] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: Debug PCIR: 00000000 at  00000048
[ 2267.914273] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: Debug PCIW: 00000040 at  00000048
[ 2267.915962] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: Turning i915 HDAC power 0
[ 2267.916250] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: Current DSP power state: D0I0
[ 2267.916256] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: fw_state change: 0 -> 1
[ 2267.916262] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: loading firmware
[ 2267.916266] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: fw_state change: 1 -> 2
[ 2267.916301] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: Debug PCIR: 00000002 at  00000048
[ 2267.916327] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: Debug PCIW: 00000000 at  00000048
[ 2267.916375] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: Debug PCIR: 00000000 at  00000044
[ 2267.916410] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: Debug PCIW: 00000004 at  00000044
[ 2267.916415] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: booting DSP firmware
[ 2267.917092] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: FW Poll Status: reg[0x160]=0x40000 successful
[ 2267.917454] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: FW Poll Status: reg[0x160]=0x40000 successful
[ 2267.917488] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: hda_dsp_stream_setup_bdl: period_bytes:0x0
[ 2267.917494] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: hda_dsp_stream_setup_bdl: periods:1
[ 2267.917814] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: Attempting iteration 0 of Core En/ROM load...
[ 2267.917843] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: DSP core(s) enabled? 0 : core_mask f
[ 2267.917892] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: FW Poll Status: reg[0x4]=0xf0f0f0f successful
[ 2267.917977] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: FW Poll Status: reg[0x4]=0xf0f0f00 successful
[ 2267.918000] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: unstall/run core: core_mask = f
[ 2267.918058] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: DSP core(s) enabled? 1 : core_mask f
[ 2267.918768] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: FW Poll Status: reg[0x4]=0xf0f0000 successful
[ 2267.918788] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: unstall/run core: core_mask = 1
[ 2267.918824] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: DSP core(s) enabled? 1 : core_mask 1
[ 2267.932773] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: FW Poll Status: reg[0xd4]=0x80000000 successful
[ 2267.932810] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: FW Poll Status: reg[0x4]=0xf0f0e0e successful
[ 2267.932829] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: FW Poll Status: reg[0x4]=0x1010e0e successful
[ 2267.932838] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: DSP core(s) enabled? 0 : core_mask e
[ 2267.932969] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: error : DSP panic!
[ 2267.932977] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: panic: dsp_oops_offset 788480 offset 788480
[ 2267.932985] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: ------------[ DSP dump start ]------------
[ 2267.932989] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: fw_state: SOF_FW_BOOT_IN_PROGRESS (2)
[ 2267.933007] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: status: fw entered - code 00000005
[ 2267.933206] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: reason: assertion failed (0xb)
[ 2267.933211] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: trace point: 0x000041c0
[ 2267.933215] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: panic at src/schedule/edf_schedule.c:98
[ 2267.933221] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: error: DSP Firmware Oops
[ 2267.933225] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: EXCCAUSE 0x0000003f EXCVADDR 0x00000000 PS       0x00060025 SAR     0x00000000
[ 2267.933233] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: EPC1     0x00000000 EPC2     0xbe02be0d EPC3     0x00000000 EPC4    0x00000000
[ 2267.933241] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: EPC5     0x00000000 EPC6     0x00000000 EPC7     0x00000000 DEPC    0x00000000
[ 2267.933246] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: EPS2     0x00060c20 EPS3     0x00000000 EPS4     0x00000000 EPS5    0x00000000
[ 2267.933252] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: EPS6     0x00000000 EPS7     0x00000000 INTENABL 0x00000000 INTERRU 0x00000222
[ 2267.933284] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: stack dump from 0xbe086140
[ 2267.933291] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: 0xbe086140: 00000000 00000000 00000000 00000000
[ 2267.933298] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: 0xbe086150: 00000000 00000000 00000000 00000000
[ 2267.933319] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: 0xbe086160: be086140 00000000 00000000 00000000
[ 2267.933328] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: 0xbe086170: 00000000 00000000 00000000 00000000
[ 2267.933335] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: 0xbe086180: 00000000 00000000 00000000 00000000
[ 2267.933342] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: 0xbe086190: 00000000 00000000 2f637273 65686373
[ 2267.933347] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: 0xbe0861a0: 656c7564 6664652f 6863735f 6c756465
[ 2267.933352] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: 0xbe0861b0: 00632e65 00000000 00000062 ffffffff
[ 2267.933356] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: ------------[ DSP dump end ]------------
[ 2268.233472] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: FW Poll Status: reg[0x80000]=0x5 timedout
[ 2268.233507] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: FW Poll Status: reg[0x4]=0x1010f0f successful
[ 2268.233713] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: FW Poll Status: reg[0x4]=0xf0f successful
[ 2268.233721] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: DSP core(s) enabled? 0 : core_mask f
[ 2268.233726] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: Attempting iteration 1 of Core En/ROM load...
[ 2268.233744] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: DSP core(s) enabled? 0 : core_mask f
[ 2268.234405] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: FW Poll Status: reg[0x4]=0xf0f0f0f successful
[ 2268.234434] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: FW Poll Status: reg[0x4]=0xf0f0f00 successful
[ 2268.234444] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: unstall/run core: core_mask = f
[ 2268.234456] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: DSP core(s) enabled? 1 : core_mask f
[ 2268.234479] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: FW Poll Status: reg[0x4]=0xf0f0000 successful
[ 2268.234487] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: unstall/run core: core_mask = 1
[ 2268.234498] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: DSP core(s) enabled? 1 : core_mask 1
[ 2268.235377] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: FW Poll Status: reg[0xd4]=0x80000000 successful
[ 2268.235405] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: FW Poll Status: reg[0x4]=0xf0f0e0e successful
[ 2268.235702] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: FW Poll Status: reg[0x4]=0x1010e0e successful
[ 2268.235709] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: DSP core(s) enabled? 0 : core_mask e
[ 2268.324348] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: FW Poll Status: reg[0x80000]=0x5000001 successful
[ 2268.341431] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: FW Poll Status: reg[0x80000]=0x5 successful
[ 2268.341460] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: FW Poll Status: reg[0x160]=0x140000 successful
[ 2268.341476] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: Firmware download successful, booting...
[ 2268.363607] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: ipc rx: 0x70000000: FW_READY
[ 2268.363620] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: ipc: DSP is ready 0x70000000 offset 0x81000
[ 2268.363627] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: fw_state change: 2 -> 5
[ 2268.363643] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: ipc rx done: 0x70000000: FW_READY
[ 2268.363708] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: firmware boot complete
[ 2268.363739] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: Debug PCIR: 00000000 at  00000048
[ 2268.363760] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: Debug PCIW: 00000002 at  00000048
[ 2268.363773] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: Debug PCIR: 00000004 at  00000044
[ 2268.363782] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: Debug PCIW: 00000000 at  00000044
[ 2268.363809] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: FW Poll Status: reg[0x80]=0x40000 successful
[ 2268.363847] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: FW Poll Status: reg[0x80]=0x40000 successful
[ 2268.363857] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: hda_dsp_stream_setup_bdl: period_bytes:0x0
[ 2268.363862] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: hda_dsp_stream_setup_bdl: periods:1
[ 2268.363891] kernel: sof-audio-pci-intel-cnl 0000:00:1f.3: snd_sof_init_trace_ipc: stream_tag: 1
```
</details>

We have a DSP panic! during the **DSP boot retry** protection and we indeed 'recover' from the DSP panic since the boot is timing out and we reboot the DSP for another try.

In short: from DSP panic there is no way to recover, unless we recover from it.
The more important thing is that if there is a chance to recover from the unrecoverable we should not set the fw_state to crashed and we should not alarm the user (unless we are instructed to do so) by an error print.